### PR TITLE
Dockerization: first draft

### PR DESCRIPTION
**Pull request by phargogh**
Hey all,

This first draft of a PR includes docker configuration for running the various components of this application:

1. The frontend (which, when launched, is served at ~~`localhost:80`~~ `localhost:3000`)
2. The backend API (which, when launched, is served at `localhost:8000`)
3. A worker for doing geospatial things (which, when launched, is not available on a port but pings the backend)

Each of these components now has a dockerfile that contains the runtime needed for executing application code.  So a development workflow for this might be:

1. make a change
2. try it out with `docker compose up`
3. when ready, shut down the container cluster with `Ctrl+C`

The main advantages of this approach are:
1. We don't really have to think much about networking
2. Deploying this configuration appears to be a straightforward next step
3. Whenever we want to preserve the state of our containers and start keeping those around, it's pretty straightforward to do that.

## Next Steps

1. Add configuration to enable running this docker configuration on our cloud VM
2. Enable push-to-deploy so we don't have to SSH into the VM every time we want to make a change to what's deployed

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
     "eslint-plugin-jsx-a11y": "^6.5.1",
     "eslint-plugin-react": "^7.28.0",
     "eslint-plugin-react-hooks": "^4.3.0",
+    "run": "^1.4.0",
```
Comment (emlys) [inline frontend/package.json:28]: where is this used?

**Review by emlys — COMMENTED** [review]
@phargogh this looks good to me! I had a question about the worker container. As I'm understanding it the worker would `do_work` just once when its dockerfile runs and then exit. How does it work in response to server requests?

File: backend-worker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# syntax=docker/dockerfile:1
+# forcing --platform=linux/amd64 because we don't have arm builds of our conda-forge packages yet
+FROM --platform=linux/amd64 mambaorg/micromamba:0.24.0-buster
+
+ARG MAMBA_DOCKERFILE_ACTIVATE=1
```
[the remark this answers is not in the document]
Comment (phargogh) [reply backend-worker/Dockerfile:5]: Yes, but with the added benefit of being able to be set as a parameter at build time (see [docker docs](https://docs.docker.com/engine/reference/builder/#arg))

For this, I was following https://github.com/mamba-org/micromamba-docker#activating-a-conda-environment-for-run-commands, and I found that if it wasn't set, we wouldn't be able to find python in subsequent `RUN` commands.

File: backend-worker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+RUN micromamba install -c conda-forge -y pygeoprocessing natcap.invest requests flask_cors
+
+CMD /opt/conda/bin/python /opt/worker/worker.py --queue_ip 0.0.0.0 --queue_port 8000
```
[the remark this answers is not in the document]
Comment (phargogh) [reply backend-worker/Dockerfile:8]: Yep!

File: frontend/package.json (hunk 1 above)
Comment (phargogh) [reply frontend/package.json:28]: It isn't (as far as I know), but I was having some problems with running the dev server without it.

But also, maybe I'm running the dev server incorrectly, which is one of the reasons I wanted to be sure to have you and @davemfish take a look at this.

**Comment by phargogh** [comment]
>@emlys I had a question about the worker container. As I'm understanding it the worker would do_work just once when its dockerfile runs and then exit. How does it work in response to server requests?

The intent is for the worker process to be running a loop that looks like this:
```
while True:
    if server_has_work_for_me_to_do():
        do_work()
    else:
         sleep(1)
```

But you also make a good point about restarting the worker container if it fails, which is something `docker compose` is supposed to be able to handle for us.  I'll take a look at making it so.

- [ ] have docker compose restart a failed container

File: frontend/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,9 @@
+FROM node:18-bullseye
+
+# Just install the packages; the source code will be loaded in later.
+RUN mkdir /opt/frontend
+ADD ./package.json /opt/frontend/
+WORKDIR /opt/frontend
+RUN npm install --production=false
+
+CMD ./node_modules/.bin/vite --host 0.0.0.0
```
Comment (davemfish) [inline frontend/Dockerfile:9]: What about a `yarn start` instead?

File: frontend/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,9 @@
+FROM node:18-bullseye
+
+# Just install the packages; the source code will be loaded in later.
+RUN mkdir /opt/frontend
+ADD ./package.json /opt/frontend/
+WORKDIR /opt/frontend
+RUN npm install --production=false
```
Comment (davemfish) [inline frontend/Dockerfile:7]: We've been using `yarn` instead `npm`, for what it's worth

Comment (phargogh) [reply frontend/Dockerfile]: Updated!

File: frontend/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+# Just install the packages; the source code will be loaded in later.
+RUN mkdir /opt/frontend
+WORKDIR /opt/frontend
+
+ADD ./package.json /opt/frontend/
+RUN yarn install --production=false
+
+# individual files in the parent directory need to be added here
+ADD ./vite.config.js /opt/frontend/
+ADD ./index.html /opt/frontend/
```
Comment (davemfish) [inline frontend/Dockerfile:12]: We also have a `.env` file with API tokens that is not tracked by git but seems like we need to add it here.

**Review by davemfish — CHANGES_REQUESTED** [review]
Hey @phargogh I tried this out and it seems to work great for the frontend. There's one more detail though that you didn't know about...we each have a `.env` file in `frontend/` with various API tokens. Do we need to explicitly include that file in the docker build?

File: frontend/Dockerfile (hunk 6 above)
Comment (davemfish) [reply frontend/Dockerfile:12]: It's located in the same place as `index.html`, etc.

Comment (phargogh) [reply frontend/Dockerfile:12]: Sure!  I've added it in aa2fcdb.

One thing we may want to consider for the future is that we have some other way of sharing secrets.  I think `docker-compose.yml` can support this, or we could put the `.env` file in a directory and mount that into the container.  My concern is that if we add the secret to the container as we are now, that we might accidentally end up publishing the secret along with the container, so making it harder for us to accidentally do that might be wise.

Comment (phargogh) [reply frontend/Dockerfile:12]: @davemfish it appears that `docker-compose.yml` does support mounting secret files into the container at runtime!  But the catch is that it's mounted at `/run/secrets/frontend-api-keys/.env`.  If it's easier, I can see about adding an environment variable about where the `.env` file is located and we can have that set at runtime.

Added this config in 836f4e4.

Comment (davemfish) [reply frontend/Dockerfile:12]: @phargogh thanks for trying that out. We actually rely on `vite`, which relies on `dotenv` to load the `.env`. And it looks like we can set an `envDir` in `vite.config.js` to whatever we want to. So I guess that would work. It just means we can't develop the app outside the context of docker. But I suppose that's the whole point!

Comment (davemfish) [reply frontend/Dockerfile:12]: https://vitejs.dev/config/#envdir

File: backend-worker/worker.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import requests
+
+logging.basicConfig(level=logging.INFO)
+LOGGER = logging.getLogger(__name__)
+
+
+def do_work(ip, port):
+    LOGGER.info(f'Starting worker, queueing {ip}:{port}')
+    while True:
+        #response = requests.get(f'{ip}/jobsqueue/:{port}')
```
Comment (dcdenu4) [inline backend-worker/worker.py:14]: Just an FYI that I noticed with testing locally. This will need to be `f'{ip}:{port}/jobsqueue`.

**Comment by dcdenu4** [comment]
@davemfish and @phargogh
>We also have a .env file with API tokens that is not tracked by git but seems like we need to add it here.

Do we have a good method of sharing these privately so we could each test locally?

**Comment by phargogh** [comment]
> @dcdenu4 Do we have a good method of sharing these privately so we could each test locally?

Maybe google drive?  Or the private GCP bucket?

**Review by davemfish — APPROVED** [review]

**Pull request merged**